improver_tests/precipitation_type/hail_fraction/test_HailFraction.py

```diff
@@ -0,0 +1,197 @@
+        "m",
+        spatial_grid="equalarea",
+        attributes=COMMON_ATTRS,
+        standard_grid_metadata="gl_ens",
+    )
+
+    altitude_data = np.zeros((2, 2), dtype=np.float32)
+    altitude = set_up_variable_cube(
+        altitude_data,
+        "altitude",
+        "m",
+        spatial_grid="equalarea",
+        attributes=COMMON_ATTRS,
+        standard_grid_metadata="gl_ens",
+    )
+    return (
+        vertical_updraught,
+        hail_size,
+        cloud_condensation_level,
+        convective_cloud_top,
+        hail_melting_level,
+        altitude,
+    )
+
+
+@pytest.mark.parametrize("model_id_attr", (None, "mosg__model_configuration"))
+@pytest.mark.parametrize(
+    "vertical_updraught_value,hail_size_value,cloud_condensation_level_value,"
+    + "convective_cloud_top_value,hail_melting_level_value,altitude_value,expected",
+    (
+        # fmt: off
+        (1, 0.0, -10, -12, 100, 50, 0),  # No indications of hail
+        (25, 0.1, -10, -12, 100, 50, 0),  # Larger updraught, no other indications of hail
+        (5, 0.1, -2, -20, 20, 50, 0),  # Low vertical updraught prevents hail
+        (25, 0.1, -2, -20, 20, 50, 1 / 9),  # Sufficient vertical updraught, non-zero hail fraction
+        (50, 0.1, -2, -20, 20, 50, 0.25),  # Sufficient vertical updraught, non-zero hail fraction
+        (75, 0.1, -2, -20, 20, 50, 0.25),  # Large vertical updraught, non-zero hail fraction
+        (1, 3, -2, -20, 20, 50, 0.05),  # Hail size indicates non-zero hail fraction
+        (75, 0.1, -10, -20, 20, 50, 0),  # Cloud condensation level temperature prevents hail
+        (75, 0.1, -2, -10, 20, 50, 0),  # Convective cloud top temperature prevents hail
+        (75, 0.1, -2, -20, 100, 50, 0),
+        # Hail melting level prevents hail
+        # fmt: on
+    ),
+)
+def test_basic(
+    vertical_updraught_value,
+    hail_size_value,
+    cloud_condensation_level_value,
+    convective_cloud_top_value,
+    hail_melting_level_value,
+    altitude_value,
+    expected,
+    model_id_attr,
+):
+    """Test hail fraction plugin."""
+    expected_attributes = COMMON_ATTRS.copy()
+    if model_id_attr:
+        expected_attributes[model_id_attr] = setup_cubes()[0].attributes[model_id_attr]
+
+    (
+        vertical_updraught,
+        hail_size,
+        cloud_condensation_level,
+        convective_cloud_top,
+        hail_melting_level,
+        altitude,
+    ) = setup_cubes()
+
+    vertical_updraught.data = np.full_like(
+        vertical_updraught.data, vertical_updraught_value
+    )
+    hail_size.data = np.full_like(
+        hail_size.data, Unit("mm").convert(hail_size_value, "m")
+    )
+    cloud_condensation_level.data = np.full_like(
+        cloud_condensation_level.data,
+        Unit("Celsius").convert(cloud_condensation_level_value, "K"),
+    )
+    convective_cloud_top.data = np.full_like(
+        convective_cloud_top.data,
+        Unit("Celsius").convert(convective_cloud_top_value, "K"),
+    )
```

Comment:
I sort-of like the fact that you are specifying the test values in Celsius, which is easier to understand, but I don't quite like this. The target units are hard-coded, rather than being taken from the cube, so if someone wanted to change the units in the future, they would have to change it in separate places. Also, the units of the unit test value units are not specified anywhere obvious.

On reflection, I think it might be simpler to specify the unit test values in the units that the plugin expects (and that the cubes are defined in).

Reply:
I've simplified this by specifying the unit test values in the units that the plugin expects, as suggested.